bin/rose-make-docs

```diff
@@ -190,31 +186,26 @@ venv-destroy () {
     rm -rf "${VENV_PATH}"
 }
 
-json_list () {
-    # write out a bash array as a JSON list
-    local IFS=','; list="$*"; echo -n "[\"${list//,/\", \"}\"]";
-}
-
 version_file () {
     # output the dictionary {"version": ["build", ...]} in JSON format
     DOCS_DIR="$1"
 
-    echo '{'
+    ret='{'
+
     # scrape filesystem for list of rose versions which have built docs
-    mapfile -t VERSIONS < <(cd "${DOCS_DIR}"; echo "${DOC_VERSIONS[@]}")
-    for version in "${VERSIONS[@]}"; do
+    for version_dir in "${DOCS_DIR}/"*.*.*; do
+        version=$(basename "$version_dir")
         # scrape filesystem to get list of formats this version is available in
-        mapfile -t formats \
-            < <(cd "${DOCS_DIR}/${version}"; echo "${DOC_FORMATS[@]}")
-        list="    \"${version}\": $(json_list "${formats[@]}")"
-        if [[ $version == "${VERSIONS[$(( ${#VERSIONS[@]} - 1 ))]}" ]]; then
-            # JSON doesn't permit a comma after the last item in a collection
-            echo "${list}"
-        else
-            echo "${list},"
-        fi
+        ret+="\n    \"$version\": ["
+        for format_dir in "${version_dir}/"!(doctrees|*.html); do
+            format=$(basename "$format_dir")
+            ret+="$format, "
+        done
+        ret="${ret:0:-2}],"
```

Comment:
FYI `ret=${ret:0:-1}` is equivalent to backspace.

Reply:
open_mouth I was aware that this kind of thing (extracting a sub-string by indexing ultimately, I guess) could be done in Python, but not in bash. Nice!